mldsa/poly.c

```diff
@@ -400,32 +402,59 @@ void poly_challenge(poly *c, const uint8_t seed[MLDSA_CTILDEBYTES])
   shake256_finalize(&state);
   shake256_squeezeblocks(buf, 1, &state);
 
+  /* Convert the first 8 bytes of buf[] into an unsigned 64-bit value.   */
+  /* Each bit of that dictates the sign of the resulting challenge value */
   signs = 0;
   for (i = 0; i < 8; ++i)
+  __loop__(
+    assigns(i, signs)
+    invariant(i <= 8)
+  )
   {
     signs |= (uint64_t)buf[i] << 8 * i;
   }
   pos = 8;
 
-  for (i = 0; i < MLDSA_N; ++i)
-  {
-    c->coeffs[i] = 0;
-  }
+  memset(c, 0, sizeof(poly));
+
   for (i = MLDSA_N - MLDSA_TAU; i < MLDSA_N; ++i)
+  __loop__(
+    assigns(i, j, object_whole(buf), state, pos, memory_slice(c, sizeof(poly)), signs)
+    invariant(i >= MLDSA_N - MLDSA_TAU)
+    invariant(i <= MLDSA_N)
+    invariant(pos >= 1)
+    invariant(pos <= SHAKE256_RATE)
+    invariant(array_bound(c->coeffs, 0, MLDSA_N, -1, 2))
+  )
   {
     do
+    __loop__(
+      assigns(j, object_whole(buf), state, pos)
+      invariant(pos >= 1)
+      invariant(pos <= SHAKE256_RATE)
+      invariant(array_bound(c->coeffs, 0, MLDSA_N, -1, 2))
+    )
     {
       if (pos >= SHAKE256_RATE)
       {
         shake256_squeezeblocks(buf, 1, &state);
         pos = 0;
       }
+      j = buf[pos++];
+    }
+    while (j > i) /* clang-format mangles layout owing to the contract above */
+      ;
+
+    c->coeffs[i] = c->coeffs[j];
+
+    /* The least-significant bit of signs tells is if we want -1 or +1 */
+    offset = 2 * (signs & 1);
 
-      b = buf[pos++];
-    } while (b > i);
+    /* offset has value 0 or 2 here, so (1 - (int32_t) offset) has
+     * value -1 or +1 */
+    c->coeffs[j] = 1 - (int32_t)offset;
 
-    c->coeffs[i] = c->coeffs[b];
-    c->coeffs[b] = 1 - 2 * (signs & 1);
+    /* Move to the next bit of signs for next time */
```

Comment:
Curious about the deviation from the reference code here

Reply:
Agree. We should strive to deviate only if we need to, and if so, leave a comment in the commit message and a `/* Reference: ... */` comment.

Reply:
Comment added.